Add recipe to replace ManagedBean with Named annotation

evie-lau opened the pull request:
## What's changed?
New recipe to migrate `@ManagedBean` from JSF/Jakarta Faces to the recommended CDI-based approach using `@jakarta.inject.Named`

## What's your motivation?
ManagedBean is removed from EE10/Faces 4.0, and the existing [recipe](https://github.com/openrewrite/rewrite-migrate-java/blob/90713334d935cc8577dd9f7c2518480c89b0b6ac/src/main/resources/META-INF/rewrite/jakarta-faces-4.yml#L204) doesn't migrate the `@ManagedBean` annotation.
It seems using the CDI `@Named` is the general approach to replacing `@ManagedBean`.

## Anything in particular you'd like reviewers to focus on?
<!-- You can also start a discussion on particular aspects of your implementation on the files tab yourself. -->

## Anyone you would like to review specifically?
<!-- @mention them here -->

## Have you considered any alternatives or workarounds?
<!-- Any other ways to solve the problem, or ways to work around the problem. -->

## Any additional context
<!-- Any thoughts you would like to share in addition to the above. -->
This current recipe converts both javax and jakarta ManagedBean to the jakarta CDI Named annotation, making this appropriate for EE9/10 migration, but not for EE8 users wanting to switch to CDI. 
#### Should this be made more flexible to maintain either `javax` or `jakarta` based on the original code?

### Checklist
- [ ] I've added unit tests to cover both positive and negative cases
- [ ] I've read and applied the [recipe conventions and best practices](https://docs.openrewrite.org/authoring-recipes/recipe-conventions-and-best-practices)
- [ ] I've used the IntelliJ IDEA auto-formatter on affected files

timtebeek commented on src/main/java/org/openrewrite/java/migrate/jakarta/UpdateManagedBeanToNamed.java:
```diff
@@ -0,0 +1,85 @@
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new JavaIsoVisitor<ExecutionContext>() {
+            private final AnnotationMatcher MANAGED_BEAN_MATCHER_JAVAX = new AnnotationMatcher("javax.faces.bean.ManagedBean");
+            private final AnnotationMatcher MANAGED_BEAN_MATCHER_JAKARTA = new AnnotationMatcher("jakarta.faces.bean.ManagedBean");
```
Would you mind adding a quick precondition here on the use of those types? That'd make for a cheap performance improvement as it avoid visiting any compilation units unnecessarily.

timtebeek commented on src/main/java/org/openrewrite/java/migrate/jakarta/UpdateManagedBeanToNamed.java:
```diff
@@ -0,0 +1,85 @@
+            public J.Annotation visitAnnotation(J.Annotation annotation, ExecutionContext ctx) {
+                if (MANAGED_BEAN_MATCHER_JAVAX.matches(annotation) || MANAGED_BEAN_MATCHER_JAKARTA.matches(annotation)) {
+                    // Get the name from the @ManagedBean annotation
+                    String beanName;
+                    beanName = annotation.getArguments() == null ? null :
+                            annotation.getArguments().stream()
+                                    .filter(J.Assignment.class::isInstance)
+                                    .map(J.Assignment.class::cast)
+                                    .filter(arg -> arg.getVariable().toString().equals("name"))
+                                    .findFirst()
+                                    .map(arg -> arg.getAssignment().toString())
+                                    .orElse(null);
```
You might be able to get the bean name more easily with a construct similar to what's seen here:
https://github.com/openrewrite/rewrite-spring/blob/16c63561b254f72a33ac1f70926cc62d233a8d70/src/main/java/org/openrewrite/java/spring/trait/SpringBean.java#L46-L50

timtebeek commented on src/main/java/org/openrewrite/java/migrate/jakarta/UpdateManagedBeanToNamed.java:
```diff
@@ -0,0 +1,85 @@
+                    maybeRemoveImport("javax.faces.bean.ManagedBean");
+                    maybeRemoveImport("jakarta.faces.bean.ManagedBean");
+                }
+                return annotation;
```
If there's any chance of these appearing in nested annotations, then it might make sense to call `super.visitAnnotation` here.

evie-lau replied:
I don't think there's any chance of nested annotations here.

timtebeek commented on src/main/java/org/openrewrite/java/migrate/jakarta/UpdateManagedBeanToNamed.java:
```diff
@@ -0,0 +1,85 @@
+                                    .findFirst()
+                                    .map(arg -> arg.getAssignment().toString())
+                                    .orElse(null);
+                    // Replace the @ManagedBean annotation with @Named
+                    if (beanName != null) {
+                        annotation = JavaTemplate.builder("@Named(\"#{}\")")
+                                .javaParser(JavaParser.fromJavaVersion().classpathFromResources(ctx, "jakarta.inject-api-2.0.1"))
+                                .imports("jakarta.inject.Named")
+                                .build()
+                                .apply(getCursor(), annotation.getCoordinates().replace(), beanName);
+                    } else {
+                        annotation = JavaTemplate.builder("@Named")
+                                .javaParser(JavaParser.fromJavaVersion().classpathFromResources(ctx, "jakarta.inject-api-2.0.1"))
+                                .imports("jakarta.inject.Named")
+                                .build()
+                                .apply(getCursor(), annotation.getCoordinates().replace());
+                    }
+                    maybeAddImport("jakarta.inject.Named");
+                    maybeRemoveImport("javax.faces.bean.ManagedBean");
+                    maybeRemoveImport("jakarta.faces.bean.ManagedBean");
```
Slight preference for immediate returns here, unless you want to call `super` with the modified annotation below.
```suggestion
                    maybeAddImport("jakarta.inject.Named");
                    maybeRemoveImport("javax.faces.bean.ManagedBean");
                    maybeRemoveImport("jakarta.faces.bean.ManagedBean");
                    // Replace the @ManagedBean annotation with @Named
                    if (beanName != null) {
                        return JavaTemplate.builder("@Named(\"#{}\")")
                                .javaParser(JavaParser.fromJavaVersion().classpathFromResources(ctx, "jakarta.inject-api-2.0.1"))
                                .imports("jakarta.inject.Named")
                                .build()
                                .apply(getCursor(), annotation.getCoordinates().replace(), beanName);
                    } else {
                        return JavaTemplate.builder("@Named")
                                .javaParser(JavaParser.fromJavaVersion().classpathFromResources(ctx, "jakarta.inject-api-2.0.1"))
                                .imports("jakarta.inject.Named")
                                .build()
                                .apply(getCursor(), annotation.getCoordinates().replace());
                    }
```

timtebeek commented on src/main/java/org/openrewrite/java/migrate/jakarta/UpdateManagedBeanToNamed.java:
```diff
@@ -0,0 +1,85 @@
+            public J.Annotation visitAnnotation(J.Annotation annotation, ExecutionContext ctx) {
+                if (MANAGED_BEAN_MATCHER_JAVAX.matches(annotation) || MANAGED_BEAN_MATCHER_JAKARTA.matches(annotation)) {
+                    // Get the name from the @ManagedBean annotation
+                    String beanName;
+                    beanName = annotation.getArguments() == null ? null :
```
Unrelated to the below trait we can immediately assign here.
```suggestion
                    String beanName = annotation.getArguments() == null ? null :
```

timtebeek commented on src/test/java/org/openrewrite/java/migrate/jakarta/UpdateManagedBeanToNamedTest.java:
```diff
@@ -0,0 +1,84 @@
+          .recipe(new UpdateManagedBeanToNamed());
+    }
+
+    @DocumentExample
```
We don't yet support this annotation on parameterized tests. Fine to leave it out for now then.
```suggestion
```

timtebeek replied:
Alternatively two separate tests could be added, one of which carries this annotation such that it shows up in the docs.

timtebeek reviewed (commented):
Couple quick comments; great to see already! Ought to be quick to address and then merge.

timtebeek reviewed (approved):
Thanks for the quick fixes!

Should `org.openrewrite.java.migrate.jakartaUpdateManagedBeanToNamed` be added to any larger declarative migration recipe before we merge?

evie-lau commented:
Just to clarify the final decision:
- added this recipe into the Faces 4.0 migration, where ManagedBean is removed and migration is required. 
- followed up on #742 and removed the Faces Bean -> CDI bean recipe list from Faces 3.0 migration

The managed bean migration recipe can still be invoked separately for anyone on EE9 wanting to get off the deprecated package.
